File: auto_editor/exports/fcp7.py
from __future__ import annotations

import xml.etree.ElementTree as ET
from fractions import Fraction
from math import ceil
from xml.etree.ElementTree import Element

from auto_editor.ffwrapper import FileInfo
from auto_editor.timeline import Clip, v3

"""
Premiere Pro uses the Final Cut Pro 7 XML Interchange Format

See docs here:
https://developer.apple.com/library/archive/documentation/AppleApplications/Reference/FinalCutPro_XML/Elements/Elements.html

Also, Premiere itself will happily output subtlety incorrect XML files that don't
come back the way they started.
"""

DEPTH = "16"


def set_tb_ntsc(tb: Fraction) -> tuple[int, str]:
    # See chart: https://developer.apple.com/library/archive/documentation/AppleApplications/Reference/FinalCutPro_XML/FrameRate/FrameRate.html#//apple_ref/doc/uid/TP30001158-TPXREF103
    if tb == Fraction(24000, 1001):
        return 24, "TRUE"
    if tb == Fraction(30000, 1001):
        return 30, "TRUE"
    if tb == Fraction(60000, 1001):
        return 60, "TRUE"

    ctb = ceil(tb)
    if ctb not in {24, 30, 60} and ctb * Fraction(999, 1000) == tb:
        return ctb, "TRUE"

    return int(tb), "FALSE"


def speedup(speed: float) -> Element:
    fil = Element("filter")
    effect = ET.SubElement(fil, "effect")
    ET.SubElement(effect, "name").text = "Time Remap"
    ET.SubElement(effect, "effectid").text = "timeremap"

    para = ET.SubElement(effect, "parameter", authoringApp="PremierePro")
    ET.SubElement(para, "parameterid").text = "variablespeed"
    ET.SubElement(para, "name").text = "variablespeed"
    ET.SubElement(para, "valuemin").text = "0"
    ET.SubElement(para, "valuemax").text = "1"
    ET.SubElement(para, "value").text = "0"

    para2 = ET.SubElement(effect, "parameter", authoringApp="PremierePro")
    ET.SubElement(para2, "parameterid").text = "speed"
    ET.SubElement(para2, "name").text = "speed"
    ET.SubElement(para2, "valuemin").text = "-100000"
    ET.SubElement(para2, "valuemax").text = "100000"
    ET.SubElement(para2, "value").text = str(speed)

    para3 = ET.SubElement(effect, "parameter", authoringApp="PremierePro")
    ET.SubElement(para3, "parameterid").text = "frameblending"
    ET.SubElement(para3, "name").text = "frameblending"
    ET.SubElement(para3, "value").text = "FALSE"

    return fil


def media_def(
    filedef: Element, url: str, src: FileInfo, tl: v3, tb: int, ntsc: str
) -> None:
    ET.SubElement(filedef, "name").text = src.path.stem
    ET.SubElement(filedef, "pathurl").text = url

    timecode = ET.SubElement(filedef, "timecode")
    ET.SubElement(timecode, "string").text = "00:00:00:00"
    ET.SubElement(timecode, "displayformat").text = "NDF"
    rate = ET.SubElement(timecode, "rate")
    ET.SubElement(rate, "timebase").text = f"{tb}"
    ET.SubElement(rate, "ntsc").text = ntsc

    rate = ET.SubElement(filedef, "rate")
    ET.SubElement(rate, "timebase").text = f"{tb}"
    ET.SubElement(rate, "ntsc").text = ntsc

    # DaVinci Resolve needs this tag even though it's blank
    ET.SubElement(filedef, "duration").text = ""

    mediadef = ET.SubElement(filedef, "media")

    if len(src.videos) > 0:
        videodef = ET.SubElement(mediadef, "video")

        vschar = ET.SubElement(videodef, "samplecharacteristics")
        rate = ET.SubElement(vschar, "rate")
        ET.SubElement(rate, "timebase").text = f"{tb}"
        ET.SubElement(rate, "ntsc").text = ntsc
        ET.SubElement(vschar, "width").text = f"{tl.res[0]}"
        ET.SubElement(vschar, "height").text = f"{tl.res[1]}"
        ET.SubElement(vschar, "pixelaspectratio").text = "square"

    for aud in src.audios:
        audiodef = ET.SubElement(mediadef, "audio")
        aschar = ET.SubElement(audiodef, "samplecharacteristics")
        ET.SubElement(aschar, "depth").text = DEPTH
        ET.SubElement(aschar, "samplerate").text = f"{tl.sr}"
        ET.SubElement(audiodef, "channelcount").text = f"{aud.channels}"


def resolve_write_audio(audio: Element, make_filedef, tl: v3) -> None:
    for t, aclips in enumerate(tl.a):
        track = ET.SubElement(audio, "track")
        for j, aclip in enumerate(aclips):
            src = aclip.src

            _start = f"{aclip.start}"
            _end = f"{aclip.start + aclip.dur}"
            _in = f"{aclip.offset}"
            _out = f"{aclip.offset + aclip.dur}"

            if not src.videos:
                clip_item_num = j + 1
            else:
                clip_item_num = len(aclips) + 1 + j

            clipitem = ET.SubElement(track, "clipitem", id=f"clipitem-{clip_item_num}")
            ET.SubElement(clipitem, "name").text = src.path.stem
            ET.SubElement(clipitem, "start").text = _start
            ET.SubElement(clipitem, "end").text = _end
            ET.SubElement(clipitem, "enabled").text = "TRUE"
            ET.SubElement(clipitem, "in").text = _in
            ET.SubElement(clipitem, "out").text = _out

            make_filedef(clipitem, aclip.src)

            sourcetrack = ET.SubElement(clipitem, "sourcetrack")
            ET.SubElement(sourcetrack, "mediatype").text = "audio"
            ET.SubElement(sourcetrack, "trackindex").text = f"{t + 1}"

            if src.videos:
                link = ET.SubElement(clipitem, "link")
                ET.SubElement(link, "linkclipref").text = f"clipitem-{j + 1}"
                ET.SubElement(link, "mediatype").text = "video"
                link = ET.SubElement(clipitem, "link")
                ET.SubElement(link, "linkclipref").text = f"clipitem-{clip_item_num}"

            if aclip.speed != 1:
                clipitem.append(speedup(aclip.speed * 100))


def premiere_write_audio(audio: Element, make_filedef, tl: v3) -> None:
    ET.SubElement(audio, "numOutputChannels").text = "2"
    aformat = ET.SubElement(audio, "format")
    aschar = ET.SubElement(aformat, "samplecharacteristics")
    ET.SubElement(aschar, "depth").text = DEPTH
    ET.SubElement(aschar, "samplerate").text = f"{tl.sr}"

    has_video = tl.v and tl.v[0]
    t = 0
    for aclips in tl.a:
        for channelcount in range(0, 2):  # Because "stereo" is hardcoded.
            t += 1
            track = ET.Element(
                "track",
                currentExplodedTrackIndex=f"{channelcount}",
                totalExplodedTrackCount="2",  # Because "stereo" is hardcoded.
                premiereTrackType="Stereo",
            )

            if has_video:
                ET.SubElement(track, "outputchannelindex").text = f"{channelcount + 1}"

            for j, aclip in enumerate(aclips):
                src = aclip.src

                _start = f"{aclip.start}"
                _end = f"{aclip.start + aclip.dur}"
                _in = f"{aclip.offset}"
                _out = f"{aclip.offset + aclip.dur}"

                if not has_video:
                    clip_item_num = j + 1
                else:
                    clip_item_num = len(aclips) + 1 + j + (t * len(aclips))

                clipitem = ET.SubElement(
                    track,
                    "clipitem",
                    id=f"clipitem-{clip_item_num}",
                    premiereChannelType="stereo",
                )
                ET.SubElement(clipitem, "name").text = src.path.stem
                ET.SubElement(clipitem, "enabled").text = "TRUE"
                ET.SubElement(clipitem, "start").text = _start
                ET.SubElement(clipitem, "end").text = _end
                ET.SubElement(clipitem, "in").text = _in
                ET.SubElement(clipitem, "out").text = _out

                make_filedef(clipitem, aclip.src)

                sourcetrack = ET.SubElement(clipitem, "sourcetrack")
                ET.SubElement(sourcetrack, "mediatype").text = "audio"
                ET.SubElement(sourcetrack, "trackindex").text = f"{t}"
                labels = ET.SubElement(clipitem, "labels")
                ET.SubElement(labels, "label2").text = "Iris"

                if aclip.speed != 1:
                    clipitem.append(speedup(aclip.speed * 100))

            audio.append(track)


def fcp7_write_xml(name: str, output: str, resolve: bool, tl: v3) -> None:
    width, height = tl.res
    timebase, ntsc = set_tb_ntsc(tl.tb)

    src_to_url: dict[FileInfo, str] = {}
    src_to_id: dict[FileInfo, str] = {}

    file_defs: set[str] = set()  # Contains urls

    for src in set(tl.sources):
        the_id = f"file-{len(src_to_id) + 1}"
        src_to_url[src] = f"{src.path.resolve()}"
        src_to_id[src] = the_id

    def make_filedef(clipitem: Element, src: FileInfo) -> None:
        pathurl = src_to_url[src]
        filedef = ET.SubElement(clipitem, "file", id=src_to_id[src])
        if pathurl not in file_defs:
            media_def(filedef, pathurl, src, tl, timebase, ntsc)
            file_defs.add(pathurl)

    xmeml = ET.Element("xmeml", version="5")
    if resolve:
        sequence = ET.SubElement(xmeml, "sequence")
    else:
        sequence = ET.SubElement(xmeml, "sequence", explodedTracks="true")

    ET.SubElement(sequence, "name").text = name
    ET.SubElement(sequence, "duration").text = f"{len(tl)}"
    rate = ET.SubElement(sequence, "rate")
    ET.SubElement(rate, "timebase").text = f"{timebase}"
    ET.SubElement(rate, "ntsc").text = ntsc
    media = ET.SubElement(sequence, "media")
    video = ET.SubElement(media, "video")
    vformat = ET.SubElement(video, "format")
    vschar = ET.SubElement(vformat, "samplecharacteristics")

    ET.SubElement(vschar, "width").text = f"{width}"
    ET.SubElement(vschar, "height").text = f"{height}"
    ET.SubElement(vschar, "pixelaspectratio").text = "square"

    rate = ET.SubElement(vschar, "rate")
    ET.SubElement(rate, "timebase").text = f"{timebase}"
    ET.SubElement(rate, "ntsc").text = ntsc

    if len(tl.v) > 0 and len(tl.v[0]) > 0:
        track = ET.SubElement(video, "track")

        for j, clip in enumerate(tl.v[0]):
            assert isinstance(clip, Clip)

            _start = f"{clip.start}"
            _end = f"{clip.start + clip.dur}"
            _in = f"{clip.offset}"
            _out = f"{clip.offset + clip.dur}"

            this_clipid = f"clipitem-{j + 1}"
            clipitem = ET.SubElement(track, "clipitem", id=this_clipid)
            ET.SubElement(clipitem, "name").text = src.path.stem
            ET.SubElement(clipitem, "enabled").text = "TRUE"
            ET.SubElement(clipitem, "start").text = _start
            ET.SubElement(clipitem, "end").text = _end
            ET.SubElement(clipitem, "in").text = _in
            ET.SubElement(clipitem, "out").text = _out

            make_filedef(clipitem, clip.src)

            ET.SubElement(clipitem, "compositemode").text = "normal"
            if clip.speed != 1:
                clipitem.append(speedup(clip.speed * 100))

            if resolve:
                link = ET.SubElement(clipitem, "link")
                ET.SubElement(link, "linkclipref").text = this_clipid
                link = ET.SubElement(clipitem, "link")
                ET.SubElement(
                    link, "linkclipref"
                ).text = f"clipitem-{(len(tl.v[0])) + j + 1}"
                continue

            for i in range(1 + len(src.audios) * 2):  # `2` because stereo.
                link = ET.SubElement(clipitem, "link")
                ET.SubElement(
                    link, "linkclipref"
                ).text = f"clipitem-{(i * (len(tl.v[0]))) + j + 1}"
                ET.SubElement(link, "mediatype").text = "video" if i == 0 else "audio"
                ET.SubElement(link, "trackindex").text = f"{max(i, 1)}"
                ET.SubElement(link, "clipindex").text = f"{j + 1}"

    # Audio definitions and clips
    audio = ET.SubElement(media, "audio")
    if resolve:
        resolve_write_audio(audio, make_filedef, tl)
    else:
        premiere_write_audio(audio, make_filedef, tl)

    tree = ET.ElementTree(xmeml)
    ET.indent(tree, space="  ", level=0)
    if output == "-":
        print(ET.tostring(xmeml, encoding="unicode"))
    else:
        tree.write(output, xml_declaration=True, encoding="utf-8")

File: src/exports/fcp7.nim
import std/[os, sets, strformat, tables, xmltree]
from std/math import ceil, round
when defined(windows):
  import std/strutils

import ../[action, ffmpeg, media, timeline]
import ../util/[fun, rational]

#[
Premiere Pro uses the Final Cut Pro 7 XML Interchange Format

See docs here:
https://developer.apple.com/library/archive/documentation/AppleApplications/Reference/FinalCutPro_XML/Elements/Elements.html

Also, Premiere itself will happily output subtlety incorrect XML files that don't
come back the way they started.
]#

const DEPTH = "16"


func setTbNtsc(tb: AVRational): (int64, string) =
  # See chart: https://developer.apple.com/library/archive/documentation/AppleApplications/Reference/FinalCutPro_XML/FrameRate/FrameRate.html#//apple_ref/doc/uid/TP30001158-TPXREF103
  if tb == AVRational(num: 24000, den: 1001):
    return (24'i64, "TRUE")
  if tb == AVRational(num: 30000, den: 1001):
    return (30'i64, "TRUE")
  if tb == AVRational(num: 60000, den: 1001):
    return (60'i64, "TRUE")

  var ctb = int64(ceil(tb.float64))
  if ctb notin [24'i64, 30, 60] and tb == ctb * AVRational(num: 999, den: 1000):
    return (ctb, "TRUE")

  return (tb.int64, "FALSE")


func elem(tag, text: string): XmlNode =
  let e = newElement(tag)
  e.add newText(text)
  return e

func param(id, name, value: string, min = "", max = ""): XmlNode =
  let p = <>parameter(authoringApp = "PremierePro")
  p.add elem("parameterid", id)
  p.add elem("name", name)
  if min != "": p.add elem("valuemin", min)
  if max != "": p.add elem("valuemax", max)
  p.add elem("value", value)
  return p

func speedup(speed: float): XmlNode =
  let fil = newElement("filter")
  let effect = newElement("effect")
  effect.add elem("name", "Time Remap")
  effect.add elem("effectid", "timeremap")
  effect.add param("variablespeed", "variablespeed", "0", "0", "1")
  effect.add param("speed", "speed", $speed, "-100000", "100000")
  effect.add param("frameblending", "frameblending", "FALSE")
  fil.add effect

  return fil


proc mediaDef(filedef: XmlNode, url: string, mi: MediaInfo, tl: v3, tb: int64,
    ntsc: string, durationStr: string) =
  filedef.add elem("name", agSplitFile(mi.path).name)
  filedef.add elem("pathurl", url)

  let timecode = newElement("timecode")
  timecode.add elem("string", "00:00:00:00")
  timecode.add elem("displayformat", "NDF")
  let rate1 = newElement("rate")
  rate1.add elem("timebase", $tb)
  rate1.add elem("ntsc", ntsc)
  timecode.add rate1
  filedef.add timecode

  let rate2 = newElement("rate")
  rate2.add elem("timebase", $tb)
  rate2.add elem("ntsc", ntsc)
  filedef.add rate2

  # Blank for Resolve (which needs the tag present but empty) and for real video
  # (Premiere reads the length from the media); a frame count for still images so
  # Premiere holds them instead of falling back to its huge default still length.
  filedef.add elem("duration", durationStr)

  let mediadef = newElement("media")

  if mi.v.len > 0:
    let videodef = newElement("video")
    let vschar = newElement("samplecharacteristics")
    let rate3 = newElement("rate")
    rate3.add elem("timebase", $tb)
    rate3.add elem("ntsc", ntsc)
    vschar.add rate3
    # The file def must describe the source's own frame size, not the sequence's;
    # an overlay still/clip is often a different size than the timeline.
    vschar.add elem("width", $mi.v[0].width)
    vschar.add elem("height", $mi.v[0].height)
    vschar.add elem("pixelaspectratio", "square")
    videodef.add vschar
    mediadef.add videodef

  for aud in mi.a:
    let audiodef = newElement("audio")
    let aschar = newElement("samplecharacteristics")
    aschar.add elem("depth", DEPTH)
    aschar.add elem("samplerate", $tl.sr)
    audiodef.add aschar
    audiodef.add elem("channelcount", $aud.channels)
    mediadef.add audiodef

  filedef.add mediadef

proc resolveWriteAudio(audio: XmlNode, makeFiledef: proc(clipitem: XmlNode,
    mi: MediaInfo), tl: v3, ptrToMi: Table[ptr string, MediaInfo]) =
  for t, alayer in tl.a.pairs:
    let track = newElement("track")
    for j, aclip in alayer.pairs:
      let mi = ptrToMi[aclip.src]

      let startVal = $aclip.start
      let endVal = $(aclip.start + aclip.dur)
      let inVal = $aclip.offset
      let outVal = $(aclip.offset + aclip.dur)

      let clipItemNum = if mi.v.len == 0: j + 1 else: alayer.len + 1 + j

      let clipitem = newElement("clipitem")
      clipitem.attrs = {"id": &"clipitem-{clipItemNum}"}.toXmlAttributes
      clipitem.add elem("name", agSplitFile(mi.path).name)
      clipitem.add elem("start", startVal)
      clipitem.add elem("end", endVal)
      clipitem.add elem("enabled", "TRUE")
      clipitem.add elem("in", inVal)
      clipitem.add elem("out", outVal)

      makeFiledef(clipitem, mi)

      let sourcetrack = newElement("sourcetrack")
      sourcetrack.add elem("mediatype", "audio")
      sourcetrack.add elem("trackindex", $(t + 1))
      clipitem.add sourcetrack

      if mi.v.len > 0:
        let link1 = newElement("link")
        link1.add elem("linkclipref", &"clipitem-{j + 1}")
        link1.add elem("mediatype", "video")
        clipitem.add link1
        let link2 = newElement("link")
        link2.add elem("linkclipref", &"clipitem-{clipItemNum}")
        clipitem.add link2

      let effectGroup = tl.effects[aclip.effects]
      for effect in effectGroup:
        if effect.kind in [actSpeed, actVarispeed]:
          clipitem.add speedup(effect.val * 100)
          break

      track.add clipitem
    audio.add track

type AudioTrackInfo = object
  layerIdx: int         # Index into tl.a
  isStereo: bool        # Stereo source -> exploded track-pair; else mono track
  explodedIndex: int    # currentExplodedTrackIndex (0 or 1); 0 for mono
  sourceTrackIndex: int # sourcetrack/trackindex into the source file
  firstClipId: int      # clipitem id of this track's first clip

proc planPremiereAudio(tl: v3, ptrToMi: Table[ptr string, MediaInfo],
    firstAudioId: int): seq[AudioTrackInfo] =
  ## Lay out the audio tracks. A stereo (or higher) source is split into two
  ## exploded stereo tracks; a mono source becomes a single mono track. The
  ## exploded structure can't represent mono: Premiere derives each channel
  ## from currentExplodedTrackIndex, so a mono source's second exploded track
  ## would read a non-existent channel and play silent on one side.
  var nextId = firstAudioId
  var channelOffset = 0
  for layerIdx, alayer in tl.a:
    var channels = 1
    if alayer.len > 0:
      let mi = ptrToMi[alayer[0].src]
      let st = alayer[0].stream.int
      if st >= 0 and st < mi.a.len:
        channels = max(1, mi.a[st].channels.int)
    let isStereo = channels >= 2
    for explodedIndex in 0 ..< (if isStereo: 2 else: 1):
      result.add AudioTrackInfo(
        layerIdx: layerIdx,
        isStereo: isStereo,
        explodedIndex: explodedIndex,
        sourceTrackIndex: channelOffset + explodedIndex + 1,
        firstClipId: nextId,
      )
      nextId += alayer.len
    channelOffset += channels

proc premiereWriteAudio(audio: XmlNode, makeFiledef: proc(clipitem: XmlNode,
    mi: MediaInfo), tl: v3, ptrToMi: Table[ptr string, MediaInfo],
    audioPlan: seq[AudioTrackInfo]) =
  audio.add elem("numOutputChannels", "2")
  let aformat = newElement("format")
  let aschar = newElement("samplecharacteristics")
  aschar.add elem("depth", DEPTH)
  aschar.add elem("samplerate", $tl.sr)
  aformat.add aschar
  audio.add aformat

  let hasVideo = tl.v.len > 0 and tl.v[0].len > 0
  for atrack in audioPlan:
    let alayer = tl.a[atrack.layerIdx]
    let track = newElement("track")
    track.attrs = {
      "currentExplodedTrackIndex": $atrack.explodedIndex,
      "totalExplodedTrackCount": (if atrack.isStereo: "2" else: "1"),
      "premiereTrackType": (if atrack.isStereo: "Stereo" else: "Mono")
    }.toXmlAttributes

    if hasVideo and atrack.isStereo:
      track.add elem("outputchannelindex", $(atrack.explodedIndex + 1))

    for j, aclip in alayer.pairs:
      let src = ptrToMi[aclip.src]

      let startVal = $aclip.start
      let endVal = $(aclip.start + aclip.dur)
      let inVal = $aclip.offset
      let outVal = $(aclip.offset + aclip.dur)

      let clipitem = newElement("clipitem")
      clipitem.attrs = {
        "id": &"clipitem-{atrack.firstClipId + j}",
        "premiereChannelType": (if atrack.isStereo: "stereo" else: "mono")
      }.toXmlAttributes
      clipitem.add elem("name", agSplitFile(src.path).name)
      clipitem.add elem("enabled", "TRUE")
      clipitem.add elem("start", startVal)
      clipitem.add elem("end", endVal)
      clipitem.add elem("in", inVal)
      clipitem.add elem("out", outVal)

      makeFiledef(clipitem, src)

      let sourcetrack = newElement("sourcetrack")
      sourcetrack.add elem("mediatype", "audio")
      sourcetrack.add elem("trackindex", $atrack.sourceTrackIndex)
      clipitem.add sourcetrack

      let labels = newElement("labels")
      labels.add elem("label2", "Iris")
      clipitem.add labels

      let effectGroup = tl.effects[aclip.effects]
      for effect in effectGroup:
        if effect.kind in [actSpeed, actVarispeed]:
          clipitem.add speedup(effect.val * 100)
          break

      track.add clipitem
    audio.add track

proc handlePath(src: string): string =
  let absPath = src.absolutePath()
  when defined(windows):
    absPath.replace('\\', '/')
  else:
    absPath

func isStillImage(mi: MediaInfo, tb: AVRational): bool =
  ## A source with a video stream but essentially no duration (0 or 1 frame) is a
  ## still image. Premiere holds these for a fixed length; given a blank
  ## `<duration>` it ignores the clip's in/out and falls back to a ~12h default.
  mi.v.len > 0 and int64(round(mi.duration * tb.float64)) <= 1

proc fcp7WriteXml*(name, output: string, resolve: bool, tl: v3) =
  let (width, height) = tl.res
  let (timebase, ntsc) = setTbNtsc(tl.tb)

  var miToUrl = initTable[MediaInfo, string]()
  var miToId = initTable[MediaInfo, string]()
  var ptrToMi = initTable[ptr string, MediaInfo]() # Cache ptr -> MediaInfo lookup
  var fileDefs = initHashSet[string]() # Contains urls

  var idCounter = 0
  for ptrSrc in tl.uniqueSources:
    idCounter += 1
    let src = ptrSrc[]
    let mi = initMediaInfo(src)
    miToUrl[mi] = handlePath(src)
    miToId[mi] = &"file-{idCounter}"
    ptrToMi[ptrSrc] = mi

  proc makeFiledef(clipitem: XmlNode, mi: MediaInfo) =
    let pathurl = miToUrl[mi]
    let filedef = <>file(id = miToId[mi])
    if pathurl notin fileDefs:
      let durationStr =
        if resolve: ""                              # Resolve wants it blank
        elif isStillImage(mi, tl.tb): $tl.len       # hold the still long enough
        else: ""                                    # Premiere reads media length
      mediaDef(filedef, pathurl, mi, tl, timebase, ntsc, durationStr)
      fileDefs.incl(pathurl)
    clipitem.add filedef

  let hasVideo = tl.v.len > 0 and tl.v[0].len > 0

  # clipitem ids run V1, V2, ... then audio. Overlay tracks (v[1..]) used to be
  # dropped entirely; emitting them means audio ids must start past every video
  # clip, not just v[0]'s.
  var videoTrackStart = newSeq[int](tl.v.len)
  var nextVideoId = 1
  for k in 0 ..< tl.v.len:
    videoTrackStart[k] = nextVideoId
    nextVideoId += tl.v[k].len
  let firstAudioId = (if hasVideo: nextVideoId else: 1)
  let audioPlan = (if resolve: @[]
                   else: planPremiereAudio(tl, ptrToMi, firstAudioId))

  let xmeml = <>xmeml(version = "5")
  let sequence = (if resolve: <>sequence() else: <>sequence(explodedTracks = "true"))

  sequence.add elem("name", name)
  sequence.add elem("duration", $tl.len)
  let rate1 = newElement("rate")
  rate1.add elem("timebase", $timebase)
  rate1.add elem("ntsc", ntsc)
  sequence.add rate1

  let media = newElement("media")
  let video = newElement("video")
  let vformat = newElement("format")
  let vschar = newElement("samplecharacteristics")

  vschar.add elem("width", $width)
  vschar.add elem("height", $height)
  vschar.add elem("pixelaspectratio", "square")

  let rate2 = newElement("rate")
  rate2.add elem("timebase", $timebase)
  rate2.add elem("ntsc", ntsc)
  vschar.add rate2
  vformat.add vschar
  video.add vformat

  if hasVideo:
    for vIdx, vlayer in tl.v:
      # resolve-fcp7's link/id scheme is wired for a single video track; only
      # Premiere gets the overlay tracks (v[1..]).
      if resolve and vIdx > 0:
        break
      if vlayer.len == 0:
        continue
      let track = newElement("track")
      let trackStart = videoTrackStart[vIdx]

      for j, clip in vlayer.pairs:
        # An audio-only timeline that gained `add:` overlays has a synthesized
        # base track with no source file (just a render canvas); there is nothing
        # to write for Premiere, so skip it (and drop a track left empty).
        if clip.src == nil:
          continue

        let mi = ptrToMi[clip.src]
        let still = isStillImage(mi, tl.tb)
        let startVal = $clip.start
        let endVal = $(clip.start + clip.dur)
        # A still is the same frame at every offset, so read `dur` frames from 0;
        # real media keeps its true source in/out.
        let inVal = (if still: "0" else: $clip.offset)
        let outVal = (if still: $clip.dur else: $(clip.offset + clip.dur))

        let thisClipid = &"clipitem-{trackStart + j}"
        let clipitem = <>clipitem(id = thisClipid)
        clipitem.add elem("name", agSplitFile(clip.src[]).name)
        clipitem.add elem("enabled", "TRUE")
        clipitem.add elem("start", startVal)
        clipitem.add elem("end", endVal)
        clipitem.add elem("in", inVal)
        clipitem.add elem("out", outVal)

        makeFiledef(clipitem, mi)

        clipitem.add elem("compositemode", "normal")

        let effectGroup = tl.effects[clip.effects]
        for effect in effectGroup:
          if effect.kind in [actSpeed, actVarispeed]:
            clipitem.add speedup(effect.val * 100)
            break

        if resolve:
          let link1 = newElement("link")
          link1.add elem("linkclipref", thisClipid)
          clipitem.add link1
          let link2 = newElement("link")
          link2.add elem("linkclipref", &"clipitem-{tl.v[0].len + j + 1}")
          clipitem.add link2
        elif vIdx == 0:
          # Base track: link the clip to itself and to its aligned audio clips so
          # they move together. Overlay tracks (v[1..]) carry no audio, so they
          # stay standalone rather than forming a one-member self-link group.
          let vlink = newElement("link")
          vlink.add elem("linkclipref", thisClipid)
          vlink.add elem("mediatype", "video")
          vlink.add elem("trackindex", "1")
          vlink.add elem("clipindex", $(j + 1))
          clipitem.add vlink
          for k, atrack in audioPlan:
            if j >= tl.a[atrack.layerIdx].len: continue
            let link = newElement("link")
            link.add elem("linkclipref", &"clipitem-{atrack.firstClipId + j}")
            link.add elem("mediatype", "audio")
            link.add elem("trackindex", $(k + 1))
            link.add elem("clipindex", $(j + 1))
            clipitem.add link

        track.add clipitem
      if track.len > 0:
        video.add track

  media.add video

  # Audio definitions and clips
  let audio = newElement("audio")
  if resolve:
    resolveWriteAudio(audio, makeFiledef, tl, ptrToMi)
  else:
    premiereWriteAudio(audio, makeFiledef, tl, ptrToMi, audioPlan)

  media.add audio
  sequence.add media
  xmeml.add sequence

  if output == "-":
    echo $xmeml
  else:
    let xmlStr = "<?xml version='1.0' encoding='utf-8'?>\n" & $xmeml
    writeFile(output, xmlStr)
